riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+                */
+            }
+        }
+    }
+
+    let submachines_init = runtime.submachines_init();
+    let bootloader_and_shutdown_routine_lines = if continuations {
+        let bootloader_and_shutdown_routine =
+            bootloader_and_shutdown_routine(field, &submachines_init);
+        log::debug!("Adding Bootloader:\n{}", bootloader_and_shutdown_routine);
+        bootloader_and_shutdown_routine
+            .split('\n')
+            .map(|l| l.to_string())
+            .collect::<Vec<_>>()
+    } else {
+        submachines_init
+    };
+
+    let mut statements: Vec<String> = program
+        .take_source_files_info()
+        .map(
+            |SourceFileInfo {
+                 id,
+                 dir,
+                 file: name,
+             }| { format!(".debug file {id} {} {};", quote(dir), quote(name)) },
+        )
+        .chain(bootloader_and_shutdown_routine_lines)
+        .collect();
+    if !data_code.is_empty() {
+        statements.push("jump __data_init, 1;".to_string());
+    }
+    statements.extend([
+        "set_reg 0, 0, 0;".to_string(),
+        format!(
+            "jump {}, 1;",
+            escape_label(program.start_function().as_ref())
+        ),
+        "return;".to_string(), // This is not "riscv ret", but "return from powdr asm function".
+    ]);
+    for s in program.take_executable_statements() {
+        match s {
+            Statement::DebugLoc { file, line, col } => {
+                statements.push(format!(".debug loc {file} {line} {col};"))
+            }
+            Statement::Label(l) => statements.push(format!("{}:", escape_label(l.as_ref()))),
+            Statement::Instruction { op, args } => {
+                let processed_instr = match process_instruction(op, args) {
+                    Ok(s) => s,
+                    Err(e) => panic!("Failed to process instruction '{op}'. {e}"),
+                };
+                statements.extend(processed_instr.into_iter().map(|s| "  ".to_string() + &s))
+            }
+        }
+    }
+
+    if !data_code.is_empty() {
+        statements.extend(
+            [
+                "// This is the data initialization routine.".to_string(),
+                "__data_init:".to_string(),
+            ]
+            .into_iter()
+            .chain(data_code)
+            .chain([
+                "// This is the end of the data initialization routine.".to_string(),
+                format!("jump_dyn 1, {};", Register::from("tmp1").addr()),
+            ]),
+        );
+    }
+    statements.extend(runtime.ecall_handler());
+
+    (initial_mem, statements)
+}
+
+fn riscv_machine(
+    runtime: &Runtime,
+    preamble: &str,
+    initial_memory: Vec<String>,
+    program: Vec<String>,
+) -> String {
+    format!(
+        r#"
+{}
+use std::machines::small_field::add_sub::AddSub;
+use std::machines::small_field::arith::Arith;
+machine Main with min_degree: {}, max_degree: {} {{
+AddSub add_sub(byte2);
+Arith arith_mul(byte, byte2);
```

Comment:
Not sure, I assume because they're more fundamental and used by a bunch of basic instructions of the machine. I could split it as well, but would prefer to do that in a different PR.

Reply:
Fair point, no strong opinion, closing